src/parquet/column/levels-test.cc

```diff
@@ -0,0 +1,112 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdlib>
+#include <iostream>
+#include <sstream>
+#include <string>
+
+#include <gtest/gtest.h>
+
+#include "parquet/thrift/parquet_types.h"
+#include "parquet/column/levels.h"
+
+using std::string;
+
+namespace parquet_cpp {
+
+const int NUM_LEVELS = 1024;
+
+class TestLevels : public ::testing::Test {
+ public:
+  void SetUp() {
+    level_values_.resize(NUM_LEVELS);
+    rle_bytes_.resize(NUM_LEVELS);
+    bit_packed_bytes_.resize(NUM_LEVELS);
+    max_level_ = 2;
+    for (int i = 0; i < NUM_LEVELS; i++) {
+      level_values_[i] = (i < NUM_LEVELS/2) ? 1 : 2;
+    }
+    //leave space to write the rle length value
+    rle_encoder_.Init(parquet::Encoding::RLE, max_level_, NUM_LEVELS,
+         rle_bytes_.data() + sizeof(int32_t), sizeof(uint8_t)*NUM_LEVELS);
+
+    bit_packed_encoder_.Init(parquet::Encoding::BIT_PACKED, max_level_, NUM_LEVELS,
+         bit_packed_bytes_.data(), sizeof(uint8_t)*NUM_LEVELS);
+
+    num_rle_levels_ = rle_encoder_.EncodeMany(NUM_LEVELS, level_values_.data());
+    num_bit_packed_levels_ = bit_packed_encoder_.EncodeMany(NUM_LEVELS,
+        level_values_.data());
+
+    int32_t rle_length = rle_encoder_.getRleLength();
+    (reinterpret_cast<int32_t*>(rle_bytes_.data()))[0] = rle_length;
+  }
+
+  void TearDown() {}
+
+ protected:
+  std::vector<int16_t> level_values_;
+  int max_level_;
+  LevelEncoder rle_encoder_;
+  LevelDecoder rle_decoder_;
+  LevelEncoder bit_packed_encoder_;
+  LevelDecoder bit_packed_decoder_;
+  std::vector<uint8_t> rle_bytes_;
+  std::vector<uint8_t> bit_packed_bytes_;
+  int num_rle_levels_;
+  int num_bit_packed_levels_;
+};
+
+TEST_F(TestLevels, TestRleEncoder) {
+  ASSERT_EQ(NUM_LEVELS, num_rle_levels_);
+}
+
+TEST_F(TestLevels, TestBitPackedEncoder) {
+  ASSERT_EQ(NUM_LEVELS, num_bit_packed_levels_);
+}
+
+TEST_F(TestLevels, TestRleDecoder) {
+  level_values_.clear();
+
+  rle_decoder_.Init(parquet::Encoding::RLE, max_level_, NUM_LEVELS,
+       rle_bytes_.data());
+
+  int num_values = rle_decoder_.DecodeMany(NUM_LEVELS, level_values_.data());
+
+  ASSERT_EQ(NUM_LEVELS, num_values);
+
+  for (int i = 0; i < NUM_LEVELS; i++) {
+    ASSERT_EQ(level_values_[i], (i < NUM_LEVELS/2) ? 1 : 2);
```

Comment:
oops I missed the `level_values_` clear, sorry about that. I still think it would be better to decode into new memory to make the tests more flexible